wire-ios-data-model/Source/Core Crypto/CoreCryptoKeyProvider.swift

```diff
@@ -17,126 +17,272 @@
+        else { return }
+
+        if (try fetchCoreCryptoKey()) != nil {
+            coreCryptoKeyMigrationManager.markMigrationToScopedKeyDone()
+        } else {
+            do {
+                WireLogger.coreCrypto.info("Migrating to scoped core crypto key...", attributes: .safePublic)
+
+                // Store the unscoped key as scoped key
+                let item = CoreCryptoKeychainItem(uniqueKeyId: uniqueKeyId, userID: userID)
+                try KeychainManager.storeItem(item, value: unscopedKey)
+
+                // Mark migration as done
+                coreCryptoKeyMigrationManager.markMigrationToScopedKeyDone()
+
+            } catch {
+                throw Failure.failedToScopeKey(error)
+            }
+        }
+    }
+
+    private func rotateKeyIfNeeded(path: String) async throws {
+        guard coreCryptoKeyMigrationManager.isKeyRotationNeeded else {
+            return
+        }
 
         do {
-            return try fetchCoreCryptoKey()
+            WireLogger.coreCrypto.info("Rotating core crypto key...", attributes: .safePublic)
+            try await rotateKey(path: path)
         } catch {
-            if createIfNeeded {
-                return try createCoreCryptoKey()
-            } else {
+            throw Failure.failedToRotateKey(error)
+        }
+    }
+
+    private func rotateKey(path: String) async throws {
+
+        // Get the old key, to rotate with the new key
+        guard let oldKey = try fetchCoreCryptoKey() else {
+            throw Failure.keyNotFound
+        }
+        let oldKeyId = uniqueKeyId
+        let oldItem = CoreCryptoKeychainItem(uniqueKeyId: oldKeyId, userID: userID)
+        logRotation("fetched the old key")
+
+        // Generate a new key and save it with a new key ID
+        let newKey = try KeychainManager.generateKey(numberOfBytes: 32)
+        let newKeyId = UUID()
+        let newItem = CoreCryptoKeychainItem(uniqueKeyId: newKeyId, userID: userID)
+        try KeychainManager.storeItem(newItem, value: newKey)
+        logRotation("generated and saved the new key")
+
+        do {
+            // Update the database
+            try await coreCryptoKeyMigrationManager.updateKey(path: path, oldKey: oldKey, newKey: newKey)
+            logRotation("updated the database with the new key")
+
+            // Delete the old key
+            deleteOrMarkAsStale(item: oldItem)
+            logRotation("deleted or marked old key as stale")
+
+            // Update the unique key identifier
+            uniqueKeyId = newKeyId
+            logRotation("updated unique key identifier")
+
+            // Mark the rotation as done
+            coreCryptoKeyMigrationManager.markKeyRotationAsDone()
+            logRotation("marked key rotation as done")
+
+        } catch {
+
+            // Check if we need to clean up or rollback, otherwise throw the error
+            switch error {
+            case let CoreCryptoKeyMigrationManagerError.failedToUpdateKey(underlyingError):
```

Comment:
suggestion: you can catch this error, anything else will be rethrown automatically.

```suggestion
        } catch let CoreCryptoKeyMigrationManagerError.failedToUpdateKey(underlyingError) {
            // Check if we need to clean up or rollback, otherwise throw the error
```

Reply:
[1f0a47f](https://github.com/wireapp/wire-ios/pull/3637/commits/1f0a47fdd9917d071734691e046f356d839e7cfd)